ledger/src/blockstore_meta.rs

```diff
@@ -51,9 +51,85 @@ impl Default for ConnectedFlags {
     }
 }
 
+/// Now deprecated format for completed data indexes.
+///
+/// Replaced by [`CompletedDataIndexesV2`].
+pub type CompletedDataIndexesV1 = BTreeSet<u32>;
+/// New format for completed data indexes.
+///
+/// `BTreeSet` is inefficient for de/serialization. We
+/// are migrating to a bitvec to improve that.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+#[serde(transparent)]
+pub struct CompletedDataIndexesV2 {
+    index: BitVec<MAX_DATA_SHREDS_PER_SLOT>,
+}
+
+// API for CompletedDataIndexesV2 that mirrors BTreeSet<u32> to make migration easier.
+// This allows CompletedDataIndexesV2 to be a drop-in replacement for CompletedDataIndexesV1.
+impl CompletedDataIndexesV2 {
+    #[inline]
+    pub fn iter(&self) -> impl DoubleEndedIterator<Item = u32> + '_ {
+        self.index.iter_ones().map(|i| i as u32)
+    }
+
+    /// Only needed for V1 / V2 test compatibility.
+    ///
+    /// TODO: Remove once the migration is complete.
+    #[cfg(test)]
+    #[inline]
+    pub fn into_iter(&self) -> impl DoubleEndedIterator<Item = u32> + '_ {
+        self.iter()
+    }
+
+    #[inline]
+    pub fn insert(&mut self, index: u32) {
+        self.index.insert_unchecked(index as usize);
+    }
+
+    #[inline]
+    pub fn contains(&self, index: &u32) -> bool {
+        self.index.contains(*index as usize)
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.index.is_empty()
+    }
+
+    #[inline]
+    pub fn range<R>(&self, bounds: R) -> impl DoubleEndedIterator<Item = u32> + '_
+    where
+        R: RangeBounds<u32>,
+    {
+        let start = bounds.start_bound().map(|&b| b as usize);
+        let end = bounds.end_bound().map(|&b| b as usize);
+        self.index.range((start, end)).iter_ones().map(|i| i as u32)
+    }
+}
+
+impl FromIterator<u32> for CompletedDataIndexesV2 {
+    fn from_iter<T: IntoIterator<Item = u32>>(iter: T) -> Self {
+        let index = iter.into_iter().map(|i| i as usize).collect();
+        CompletedDataIndexesV2 { index }
+    }
+}
+
+impl From<CompletedDataIndexesV2> for CompletedDataIndexesV1 {
+    fn from(value: CompletedDataIndexesV2) -> Self {
+        value.iter().collect()
+    }
+}
+
+impl From<CompletedDataIndexesV1> for CompletedDataIndexesV2 {
+    fn from(value: CompletedDataIndexesV1) -> Self {
+        value.into_iter().collect()
+    }
+}
+
 #[derive(Clone, Debug, Default, Deserialize, Serialize, Eq, PartialEq)]
 /// The Meta column family
-pub struct SlotMeta {
+pub struct SlotMetaBase<T> {
```

Comment:
nit: Does this struct have to be `pub` in order to make `SlotMetaV1` and `SlotMetaV2` public ? We obviously can't support any type for `T` here so it'd be nice to lock down the "builder" type.

Also, I guess this compiles since `CompletedDataIndexesV2` supports all the same functions that `CompletedDataIndexesV1` does too ? For example, if we begin using some other function from `BTreeSet` that `CompletedDataIndexesV2` doesn't implement, would compilation fail ?